middle_end/flambda2/types/meet_and_join_new.ml

```diff
@@ -739,35 +677,104 @@ and meet_array_contents env (array_contents1 : TG.array_contents Or_unknown.t)
+    | No_extensions -> env_imms, env_blocks
+    | Ext { when_immediate; when_block } ->
+      ( Or_bottom.bind env_imms ~f:(fun env ->
+            TE.add_env_extension_strict env when_immediate
+              ~meet_type:(New meet_type)),
+        Or_bottom.bind env_blocks ~f:(fun env ->
+            TE.add_env_extension_strict env when_block
+              ~meet_type:(New meet_type)) )
   in
-  pairwise_disjunction_meet ~join_env_extension ~meet_type blocks_meet
-    immediates_meet env blocks1 blocks2 imms1 imms2
+  let immediates_result : _ meet_result =
+    match env_imms with
+    | Bottom -> Bottom
+    | Ok env ->
+      meet_unknown meet ~contents_is_bottom:TG.is_obviously_bottom env imms1
+        imms2
+  in
+  let blocks_result : _ meet_result =
+    match env_blocks with
+    | Bottom -> Bottom
+    | Ok env ->
+      meet_unknown meet_row_like_for_blocks
+        ~contents_is_bottom:TG.Row_like_for_blocks.is_bottom env blocks1 blocks2
+  in
+  match immediates_result, blocks_result with
+  | Bottom, Bottom -> Bottom
+  | Ok (imms_result, env), Bottom ->
+    let imms = extract_value imms_result imms1 imms2 in
+    let blocks = Or_unknown.Known TG.Row_like_for_blocks.bottom in
+    (* CR vlaviron: recover when this is actually one of the inputs *)
+    direct_return (Ok (New_result (blocks, imms, TG.No_extensions), env))
+  | Bottom, Ok (blocks_result, env) ->
+    let blocks = extract_value blocks_result blocks1 blocks2 in
+    let imms = Or_unknown.Known TG.bottom_naked_immediate in
+    (* CR vlaviron: recover when this is actually one of the inputs *)
+    direct_return (Ok (New_result (blocks, imms, TG.No_extensions), env))
+  | Ok (imms_result, env_imms), Ok (blocks_result, env_blocks) ->
+    let imms = extract_value imms_result imms1 imms2 in
+    let blocks = extract_value blocks_result blocks1 blocks2 in
+    let when_immediate = to_extension env_imms in
+    let when_block = to_extension env_blocks in
+    let extensions : TG.variant_extensions =
+      if TEE.is_empty when_immediate && TEE.is_empty when_block
+      then
+        No_extensions
+        (* CR vlaviron: If both extensions have equations in common, the join
+           below will add them to the result environment. Keeping those common
+           equations in the variant extensions then becomes redundant, but we
+           don't have an easy way to detect redundancy. *)
+      else Ext { when_immediate; when_block }
+    in
+    let join_env =
+      Join_env.create initial_env ~left_env:env_imms ~right_env:env_blocks
+    in
+    let result_extension =
+      join_env_extension join_env when_immediate when_block
+    in
+    let result_env =
+      (* Not strict, as we don't expect to be able to get bottom equations from
+         joining non-bottom ones *)
+      TE.add_env_extension initial_env result_extension
+        ~meet_type:(New meet_type)
+    in
+    (* CR vlaviron: recover when this is actually one of the inputs *)
```

Comment:
I guess the difficulty of recovering when this is one of the inputs is because of the env extensions, but it seems a bit wasteful to always lose the fact that this can be one of the inputs. Could we at least add some code to recover if it is one of the inputs if the env extensions are empty?

Reply:
I believe I have implemented the feature in the last commit (that I happened to have around on my local branch, but forgot to push...)
@Ekdohibs would you have time to review the last commit ?

Reply:
(The one named "Better handling for disjunctions", not "fmt", of course)